### PR TITLE
Move decodable lists bits as a non-template class to save flash

### PR DESCRIPTION
**Pull request by andy31415**
This looks like significant savings, since decode takes a lot of space if templated and it is not actually template-dependent. On a NRF light this looks like 6K of flash savings.

image

### CODE REVIEW

File: src/app/data-model/DecodableList.h [hunk 1]
```diff
@@ -47,30 +109,13 @@ namespace DataModel {
...
-    /*
-     * @brief
-     *
-     * This call clears the TLV reader managed by this class, so it can be reused.
-     */
-    void ClearReader() { mReader.Init(nullptr, 0); }
-
     template <typename T0 = T, std::enable_if_t<DataModel::IsFabricScoped<T0>::value, bool> = true>
     void SetFabricIndex(FabricIndex fabricIndex)
     {
```
Comment (bzbarsky-apple) [inline src/app/data-model/DecodableList.h:121]: Followup: we might get some more wins by moving bits of the iterator into a non-templated super-class too.

**Review by bzbarsky-apple — APPROVED** [review]
Thank you for fixing this stuff up!

**Review by arkq — APPROVED** [review]

File: src/app/data-model/DecodableList.h [hunk 2]
```diff
@@ -47,30 +109,13 @@ namespace DataModel {
...
-
-    /*
-     * @brief
-     *
-     * This call clears the TLV reader managed by this class, so it can be reused.
-     */
-    void ClearReader() { mReader.Init(nullptr, 0); }
-
     template <typename T0 = T, std::enable_if_t<DataModel::IsFabricScoped<T0>::value, bool> = true>
     void SetFabricIndex(FabricIndex fabricIndex)
```
[the remark this answers is not in the document]
Comment (andy31415) [reply src/app/data-model/DecodableList.h:120]: created #36264 for followups

File: src/app/data-model/DecodableList.h (hunk 1 above)
Comment (andy31415) [reply src/app/data-model/DecodableList.h:121]: created #36264 for followups

**Pull request merged**